### PR TITLE
WIP - clean up Contentful plugin

### PR DESCRIPTION
**Pull request by axe312ger**
* [x] replace cache-image.js with createRemoteFile
* [ ] use createRemoteFile instead of axios in download-with-retry.js (Do we even need this? We need to have concurrency of 100 when downloading assets)
* [ ] move gatsby-image and gatsby-plugin-image related code into their own files

### CODE REVIEW

File: packages/gatsby-source-contentful/src/extend-node-type.js [hunk 1]
```diff
@@ -676,15 +677,21 @@ exports.extendNodeType = ({ type, store, reporter }) => {
...
     } = image
 
     if (contentType.indexOf(`image/`) !== 0) {
       return null
     }
 
-    const absolutePath = await cacheImage(store, image, options, reporter)
-    const extension = path.extname(absolutePath)
+    const url = createUrl(imgUrl, options)
+    const extension = path.extname(imgUrl)
```
Comment (wardpeet) [inline packages/gatsby-source-contentful/src/extend-node-type.js:688]: Does this work for all images coming from contentful?

File: packages/gatsby-transformer-sqip/src/extend-node-type.js [hunk 2]
```diff
@@ -1,22 +1,23 @@
...
   GraphQLObjectType,
   GraphQLString,
   GraphQLInt,
   GraphQLBoolean,
 } = require(`gatsby/graphql`)
-const sharp = require(`sharp`)
-const { ensureDir } = require(`fs-extra`)
+const { queueImageResizing } = require(`gatsby-plugin-sharp`)
+const { fetchRemoteFile } = require(`gatsby-core-utils`)
+const { createUrl } = require(`gatsby-source-contentful/extend-node-type`)
```
Comment (wardpeet) [inline packages/gatsby-transformer-sqip/src/extend-node-type.js:16]: We won't want to couple this to the contentful plugin so we should copy the code into this plugin.

Comment (axe312ger) [reply packages/gatsby-transformer-sqip/src/extend-node-type.js:16]: Which code do you mean by `this`? :)

File: packages/gatsby-source-contentful/src/extend-node-type.js (hunk 1 above)
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/extend-node-type.js:688]: Will double check and tell you next week!

File: packages/gatsby-transformer-sqip/src/extend-node-type.js (hunk 2 above)
Comment (wardpeet) [reply packages/gatsby-transformer-sqip/src/extend-node-type.js:16]: I mean createUrl, we don't want gatsby-transformer-sqip to have a dependency on gatsby-source-contentful

File: packages/gatsby-source-contentful/src/extend-node-type.js (hunk 1 above)
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/extend-node-type.js]: Good catch. Actually that way was unreliable. I rewrote it to set the extension based on mime type. Now even a image with wrong file extension will be properly downloaded :)

File: packages/gatsby-transformer-sqip/src/extend-node-type.js (hunk 2 above)
Comment (axe312ger) [reply packages/gatsby-transformer-sqip/src/extend-node-type.js]: Moved the require within the resolver which is only available when you got gatsby-source-contentful enabled :)

Comment (wardpeet) [reply packages/gatsby-transformer-sqip/src/extend-node-type.js]: I didn't know we had `SqipContentful` in here, that's definitely a design flaw on our end.

**Comment by axe312ger** [comment]
Hmm somehow it now wants to connect to localhost. Will check whats going wrong here. I restart the CI now in case that was an CI (network) error.

**Comment by wardpeet** [comment]
@axe312ger is the image URL referring to https://localhost? or maybe you mocked fetchRemoteFile?

File: packages/gatsby-transformer-sqip/src/extend-node-type.js [hunk 3]
```diff
@@ -223,7 +226,17 @@ async function sqipContentful({ type, cache, store }) {
           background,
         }
 
-        const absolutePath = await cacheImage(store, asset, options)
+        const extension = mimeTypeExtensions.get(contentType)
+        const url = `https:` + createUrl(imgUrl, options)
+        const name = path.basename(fileName, extension)
+
+        console.log({ extension, url, name })
```
Comment (wardpeet) [inline packages/gatsby-transformer-sqip/src/extend-node-type.js:233]: remove console .log :)

File: packages/gatsby-source-contentful/src/extend-node-type.js [hunk 4]
```diff
@@ -712,7 +728,29 @@ exports.extendNodeType = ({ type, store, reporter }) => {
...
+        return null
+      }
+
+      // 256px should be enough to properly detect the dominant color
+      if (!options.width) {
+        options.width = 256
+      }
+
+      const extension = mimeTypeExtensions.get(contentType)
+      const url = `https:` + createUrl(imgUrl, options)
```
Comment (wardpeet) [inline packages/gatsby-source-contentful/src/extend-node-type.js:745]: Why don't we add https in the createUrl?

File: packages/gatsby-core-utils/src/fetch-remote-file.ts [hunk 5]
```diff
@@ -159,6 +159,28 @@ const requestRemoteNode = (
...
+  // Create file fetch promise and store it into cache
+  const fetchPromise = fetchRemoteFile(args)
+  fetchCache.set(args.url, fetchPromise)
+  const result = await fetchPromise
+
+  // When the file is fetched, replace the promise with the result
+  fetchCache.set(args.url, result)
+
+  return result
+}
```
Comment (wardpeet) [inline packages/gatsby-core-utils/src/fetch-remote-file.ts:182]: You can return the promise always
```suggestion
export function fetchRemoteFileWithCache(
  args: IFetchRemoteFileOptions
): Promise<string> {
  // If we are already fetching the file, return the unresolved promise
  const inFlight = fetchCache.get(args.url)
  if (inFlight) {
    return inFlight
  }

  // Create file fetch promise and store it into cache
  const fetchPromise = fetchRemoteFile(args)
  fetchCache.set(args.url, fetchPromise)

  return fetchPromise 
}
```

**Review by wardpeet — COMMENTED** [review]
Small comments, almost good to go. Is this being tested in the e2e test?

File: packages/gatsby-source-contentful/src/extend-node-type.js (hunk 4 above)
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/extend-node-type.js:745]: @wardpeet because this function is used on browser side as well. That way it works on http as well.

**Comment by axe312ger** [comment]
@wardpeet should we more the cache with the original fetchRemoteFile function? @KyleAMathews voted on merging in a conversation we had

**Comment by axe312ger** [comment]
@wardpeet yes, it is covered by e2e

**Comment by wardpeet** [comment]
@axe312ger we should move it to source-filesystem but not in this PR.

File: packages/gatsby-core-utils/src/fetch-remote-file.ts [hunk 6]
```diff
@@ -159,6 +159,24 @@ const requestRemoteNode = (
...
+    return inFlight
+  }
+
+  // Create file fetch promise and store it into cache
+  const fetchPromise = fetchRemoteFile(args)
+  fetchCache.set(args.url, fetchPromise)
+
+  return fetchPromise 
+}
+
```
Comment (wardpeet) [inline packages/gatsby-core-utils/src/fetch-remote-file.ts:179]: Yes, please move this to the original fetch

Comment (axe312ger) [reply packages/gatsby-core-utils/src/fetch-remote-file.ts:179]: Ill do so :)

**Comment by axe312ger** [comment]
@wardpeet while rewriting I figured out that base64 generation was not covered by the PR.

I rewrote those parts, now everything is using fetchRemoteFile. Additionally I had to find out that fetchRemoteFile does not retry on server errors.

As we should only request 100 assets at once from the Contentful Image API, I rewrote my old wrapper to use fetchRemoteFile, limit to 100 requests at the same time plus retry logic for server errors.

In short: `download-with-retry.js` got a rewrite to use `fetchRemoteFile` and is now called `fetch-contentful-asset.js`

File: packages/gatsby-core-utils/src/fetch-remote-file.ts [hunk 7]
```diff
@@ -159,15 +161,35 @@ const requestRemoteNode = (
...
+  const inFlight = fetchCache.get(args.url)
+  if (inFlight && !skipCache) {
+    return inFlight
+  }
+
+  // Create file fetch promise and store it into cache
+  const fetchPromise = fetchFile(args)
+  fetchCache.set(args.url, fetchPromise)
+
+  return fetchPromise
```
Comment (wardpeet) [inline packages/gatsby-core-utils/src/fetch-remote-file.ts:180]: We should remove it from the fetch cache on error so you can drop the skipCache :)

```suggestion
  return fetchPromise.catch(err => {
    fetchCache.remove(args.url);
    
    throw err;
  })
```

File: packages/gatsby-source-contentful/src/extend-node-type.js (hunk 4 above)
Comment (wardpeet) [reply packages/gatsby-source-contentful/src/extend-node-type.js:709]: This goes to contentful cdn right? Than we should force https. There is no reason to do a HTTP call ever if the server supports https.

Comment (axe312ger) [reply packages/gatsby-source-contentful/src/extend-node-type.js:709]: Ahh very true. Using http images on https sides causes issues. But this would be the other way around. 

Will change it!

File: packages/gatsby-core-utils/src/fetch-remote-file.ts [hunk 8]
```diff
@@ -161,31 +163,33 @@ const requestRemoteNode = (
...
   url,
   cache,
+  cacheDir,
   auth = {},
   httpHeaders = {},
   ext,
   name,
 }: IFetchRemoteFileOptions): Promise<string> {
-  const pluginCacheDir = cache.directory
+  const pluginCacheDir = cacheDir || cache.directory
```
[the remark this answers is not in the document]
Comment (axe312ger) [reply packages/gatsby-core-utils/src/fetch-remote-file.ts:199]: I added a comment to explain it. It is used by the GATSBY_CONTENTFUL_EXPERIMENTAL_REMOTE_CACHE flag.

File: packages/gatsby-core-utils/src/fetch-remote-file.ts (hunk 7 above)
Comment (axe312ger) [reply packages/gatsby-core-utils/src/fetch-remote-file.ts]: Great! Done!

File: packages/gatsby-core-utils/src/fetch-remote-file.ts (hunk 6 above)
Comment (axe312ger) [reply packages/gatsby-core-utils/src/fetch-remote-file.ts]: Done!

File: packages/gatsby-source-contentful/src/extend-node-type.js (hunk 4 above)
Comment (axe312ger) [reply packages/gatsby-source-contentful/src/extend-node-type.js]: Done

**Comment by axe312ger** [comment]
Do we have issues with the window temp dir? https://app.circleci.com/pipelines/github/gatsbyjs/gatsby/67858/workflows/a4b66993-8470-49a9-b442-d061d954cabc/jobs/766873?invite=true#step-107-1231

**Comment by axe312ger** [comment]
@wardpeet thank you for the releases, it works fine on my projects as the number of CI fails got reduced.

I'll set this to draft and clear up this PR from noise.
